### PR TITLE
Java: Use Java 8 map.putIfAbsent and map.computeIfAbsent #8215

### LINKED ISSUE #8215: Java: Use Java 8 map.putIfAbsent and map.computeIfAbsent
Context: Someone wants to add a key-value pair to a map if the key-value pair does not exist.

Pre-Java 8:
```java
if (!map.containsKey(key)) {
    map.put(key, value);
}
```
Java 8:
```java
// If value is independent of key
map.putIfAbsent(key, value);

// If value is dependent of key
map.computeIfAbsent(key, key -> process(key));
```

For this one there are more than 20 places. Have to fix no less than 20 to qualify for a valid PR.

### PR DESCRIPTION
**Pull request by gthd**
Fixes #8215 

**PR Checklist**

Ensure that you have:
- [x] Read and understood our PR guideline: https://github.com/TEAMMATES/teammates/blob/master/docs/process.md#step-4-submit-a-pr
  - [x] Added the issue number to the "Fixes" keyword above
  - [x] Titled the PR as specified in the abovementioned document
- [x] Made your changes on a branch other than `master` and `release`
- [x] Gone through all the changes in this PR and ensured that:
  - [x] They addressed one (and only one) issue
  - [ ] No unintended changes were made
- [x] Run and passed static analysis: `./gradlew lint` and `npm run lint`
- [ ] Added/updated tests, if changes in functionality were involved
- [ ] Added/updated documentation to public APIs (classes, methods, variables), if applicable

**Outline of Solution**

### CODE REVIEW

File: src/main/java/teammates/common/datatransfer/FeedbackSessionQuestionsBundle.java [hunk 1]
```diff
@@ -113,8 +113,9 @@ public void hideUnmoderatableQuestions() {
 
             if (!isResponseVisibleToInstructor || !isGiverVisibleToInstructor || !isRecipientVisibleToInstructor) {
                 questionsToHide.add(question);
-                questionResponseBundle.put(question, new ArrayList<FeedbackResponseAttributes>());
+
             }
+            questionResponseBundle.putIfAbsent(question, new ArrayList<>());
```
Comment (wenmogu) [inline src/main/java/teammates/common/datatransfer/FeedbackSessionQuestionsBundle.java]: Its nice that you are using diamond expression to shorten the code, but I think this change is unnecessary as we do not need to check if this question exist in the bundle or not, after all these `questions` will be removed from the `questionResponseBundle` in the next line.

File: src/main/java/teammates/common/datatransfer/InstructorPrivileges.java [hunk 2]
```diff
@@ -271,8 +271,8 @@ private void updatePrivilegesInSessionLevel(String sectionName, String sessionNa
     private void verifyExistenceOfsectionName(String sectionName) {
         if (!this.sessionLevel.containsKey(sectionName)) {
             addSectionWithDefaultPrivileges(sectionName);
-            this.sessionLevel.put(sectionName, new LinkedHashMap<String, Map<String, Boolean>>());
         }
+        this.sessionLevel.putIfAbsent(sectionName, new LinkedHashMap<>());
```
Comment (wenmogu) [inline src/main/java/teammates/common/datatransfer/InstructorPrivileges.java]: To me this `putIfAbsent` is not necessary as a few lines ago the condition `if this.sessionLevel.containsKey(sectionName)` was just checked. Therefore I think we can avoid the duplicate checking here.

File: src/main/java/teammates/logic/core/FeedbackResponsesLogic.java [hunk 3]
```diff
@@ -771,9 +771,9 @@ private void addNewResponses(
         }
         for (FeedbackResponseAttributes newResponse : newResponses) {
             if (!responses.containsKey(newResponse.getId())) {
-                responses.put(newResponse.getId(), newResponse);
                 existingResponses.add(newResponse);
             }
+            responses.putIfAbsent(newResponse.getId(), newResponse);
```
Comment (wenmogu) [inline src/main/java/teammates/logic/core/FeedbackResponsesLogic.java]: The duplicate checking from my point of view

File: src/main/java/teammates/ui/controller/AdminSearchPageAction.java [hunk 4]
```diff
@@ -87,22 +87,22 @@ private AdminSearchPageData putCourseNameIntoMap(List<StudentAttributes> student
                                                      List<InstructorAttributes> instructors,
                                                      AdminSearchPageData data) {
         for (StudentAttributes student : students) {
-            if (student.course != null && !data.courseIdToCourseNameMap.containsKey(student.course)) {
+            if (student.course != null) {
                 CourseAttributes course = logic.getCourse(student.course);
                 if (course != null) {
                     //TODO: [CourseAttribute] remove desanitization after data migration
-                    data.courseIdToCourseNameMap.put(
+                    data.courseIdToCourseNameMap.putIfAbsent(
```
Comment (wenmogu) [inline src/main/java/teammates/ui/controller/AdminSearchPageAction.java]: To me it seems that this change will incur a potential slowdown (by a little bit i guess) due to the previously avoidable `logic.getCourse(student.course)`, but the logic is clearer :)

File: src/client/java/teammates/client/scripts/StatisticsPerInstitute.java [hunk 5]
```diff
@@ -120,12 +120,11 @@ private StatsBundle generateStatsPerInstitute(
...
             String institute = getInstituteForInstructor(instructor, allAccounts);
 
-            if (!institutes.containsKey(institute)) {
-                institutes.put(institute, new HashMap<Integer, Set<String>>());
-                institutes.get(institute).put(INSTRUCTOR_INDEX, new HashSet<String>());
-                institutes.get(institute).put(STUDENT_INDEX, new HashSet<String>());
-            }
-            institutes.get(institute).get(INSTRUCTOR_INDEX).add(instructor.getEmail().toLowerCase());
+            institutes.computeIfAbsent(institute, key -> new HashMap<>()).put(INSTRUCTOR_INDEX, new HashSet<String>());
+            institutes.computeIfAbsent(institute, key -> new HashMap<>()).put(STUDENT_INDEX, new HashSet<String>());
```
[the remark this answers is not in the document]
Comment (gthd) [reply src/client/java/teammates/client/scripts/StatisticsPerInstitute.java]: Hi @wenmogu !! When you say unnecessary are they wrong or you mean that the get put does the same work, so no need to bother changing them ? I can change them back if you think so.

**Comment by gthd** [comment]
Hi @wenmogu !! When you say unnecessary are they wrong or you mean that the get put does the same work, so no need to bother changing them ? I can change them back if you think so.

File: src/client/java/teammates/client/scripts/StatisticsPerInstitute.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (wenmogu) [reply src/client/java/teammates/client/scripts/StatisticsPerInstitute.java]: They r not wrong. However, 1. It decreases the readability of the code 2. It incurs extra work when running the code as some of the checks r duplicate

[the remark this answers is not in the document]
Comment (gthd) [reply src/client/java/teammates/client/scripts/StatisticsPerInstitute.java]: Ok, I will change them back!!

**Comment by gthd** [comment]
I have created a new pull request for this issue because I deleted my previous fork. Please see https://github.com/TEAMMATES/teammates/pull/8705

**Comment by wkurniawan07** [comment]
@gthd why did you need to delete your previous fork, though?

**Comment by gthd** [comment]
Hi @wkurniawan07 Because of merge conflicts I couldn't understand how to resolve. After spending many hours to find a solution, I deleted the fork and started from scratch.

**Comment by wkurniawan07** [comment]
@gthd, it will do you more good if you actually learn how to resolve merge conflicts. You will definitely encounter it again in the future, and it's not necessarily in this project.

**Comment by gthd** [comment]
Hi @wkurniawan07 thanks for the advice. I will learn how to resolve merge conflicts.

**Pull request closed without merging**